### ISSUE TITLE
remove laminas/laminas-loader

### ISSUE THREAD

**Issue by arhimede** [laminas/laminas-http#88]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | yes

#### Summary
Since laminas/laminas-loader is abandoned, we should remove it from composer.json

**Comment by arhimede** [comment]
yeah, it is not necessary a new release , the issue will be fixed downstream

**Comment by froschdesign** [comment]
@arhimede 
> yeah, it is not necessary a new release , the issue will be fixed downstream

laminas-loader must be removed but this needs also a replacement for `HeaderLoader`:

https://github.com/laminas/laminas-http/blob/a66bfb65c698aad6ee9f10df42cb5902f2c3dec8/src/HeaderLoader.php#L3-L10

**Comment by arhimede** [comment]
@froschdesign then what should we do here, since this package is in security-only mode ?
Should we have a new release of it, still ?

**Comment by froschdesign** [comment]
@arhimede 
> …since this package is in security-only mode ?

Is not! See your own polls. 😄

**Comment by gsteel** [comment]
We voted not to abandon this one. It's integral to MVC so we can't just kill it off. I don't understand the need to deal with the abandonment of the loader library - instead we should focus on eradicating the use of this library in all dependents so that we _can_ abandon it at a future date… which is gonna take a while…

**Comment by arhimede** [comment]
> @arhimede
> 
> > …since this package is in security-only mode ?
> 
> Is not! See your own polls. 😄

I saw that , but Laminas bot closed automatically this issue because "security-only"
https://github.com/laminas/laminas-http/issues/88#issuecomment-2503557682
Do we have an issue with laminas-bot ?

**Comment by froschdesign** [comment]
> Do we have an issue with laminas-bot ?

No, no problem, but as long as the workflow is still active, the message will keep coming back and issues will be closed: https://github.com/laminas/laminas-http/blob/2.21.x/.github/workflows/auto-close.yml

**Comment by arhimede** [comment]
> > Do we have an issue with laminas-bot ?
> 
> No, no problem, but as long as the workflow is still active, the message will keep coming back and issues will be closed: https://github.com/laminas/laminas-http/blob/2.21.x/.github/workflows/auto-close.yml

There is an active auto-close Action. 
https://github.com/laminas/laminas-http/blob/2.21.x/.github/workflows/auto-close.yml
Should we remove it maybe ?

**Comment by froschdesign** [comment]
> Should we remove it maybe ?

Yes, it needs to be removed.

**Comment by arhimede** [comment]
ok, @alexmerlin Let's see what we can do here, i re-opened the ticket

**Comment by alexmerlin** [comment]
I analyzed the code and found that there are 2 places where laminas-loader is being used:

* `src/HeaderLoader.php` - extends `Laminas\Loader\PluginClassLoader` which is an implementation of the `Laminas\Loader\PluginClassLocator` interface
* `src/Headers.php` - contains `$pluginClassLoader` which is of type `Laminas\Loader\PluginClassLocator`

The simplest solution to allow the removal of `laminas-loader` would be to copy `PluginClassLoader` and `PluginClassLocator` inside of `laminas-http`.
But, being that `Headers->setPluginClassLoader()` is called from outside of `laminas-http`, not sure if the caller could be set up to call `setPluginClassLoader` with the custom `HeaderLoader`.

@gsteel What do you think?

**Comment by gsteel** [comment]
Changing the inheritance hierarchy of `HeaderLoader` will be a BC break, and the reason `loader` is abandoned is because we have composer now.

Seeing as the inheritance changes would be a BC break, we might as well put the effort in to refactoring `HeaderLoader` instead so that it doesn't rely on loader at all.

I'm guessing, it turns `content-type` into `Laminas\Http\Header\ContentType` or something like that… and allows users to provide custom header implementations for `x-my-header` to `My\Header\ClassName` - I would start by re-implementing that feature so that it doesn't require any external dependencies.

Releasing a major just to get rid of loader seems excessive… especially on a lib that is destined to be abandoned itself.

Could we focus effort of ridding other libs of `laminas-http` instead?

I may be off the mark - it's been a long time since I looked at `laminas-http`…

**Comment by froschdesign** [comment]
@gsteel 
> …especially on a lib that is destined to be abandoned itself.

The vote says something different here too.

**Comment by SuperDJ** [comment]
Should be closed by https://github.com/laminas/laminas-http/pull/110
